dev-docs/RFCs/v6.2/mapbox-custom-layer-rfc.md

```diff
@@ -0,0 +1,191 @@
+
+
+## Proposed MapboxGL API Integration
+
+This case is fairly simple: a `MapboxLayer` class that can be used as a custom mapbox layer provides the plumbing and enables mapbox users to use deck.gl "layer packs".
+
+The adapter class transparently creates and manages as `Deck` instance under the hood.
+
+```js
+import Deck, {ScatterplotLayer} from '@deck.gl/core';
+import {MapboxLayer} from '@deck.gl/mapbox';
+
+const map = new mapboxgl.Map({
+  container: 'map',
+  style: 'mapbox://styles/mapbox/light-v9',
+  center: [-74.012, 40.705],
+  zoom: 15.5,
+  bearing: -20,
+  pitch: 45
+});
+
+map.on('load', () => {
+  map.addLayer(mapboxBuildingLayer);
+  map.addLayer(deckPoiLayer, getFirstTextLayerId(map.getStyle()));
+  map.addLayer(deckRouteLayer);
+});
+```
+
+
+## Proposed DeckGL API Integration
+
+This case is a bit harder since it both requires postponing deck.gl initialization.
+
+### Proposal 1: Callbacks
+
+```js
+import Deck, {ScatterplotLayer} from '@deck.gl/core';
+import {MapboxLayer} from '@deck.gl/mapbox';
+
+const map = new mapboxgl.Map({...});
+
+const deck = new Deck({
+  width: '100%',
+  height: '100%',
+  longitude: -122.4,
+  latitude: 37.8,
+  layers: [
+    new MapboxLayer({
+      id: 'my-scatterplot',
+      insertBefore: 'admin-1-labels',
+      type: ScatterplotLayer,
+      ...
+    })
+  ],
+  onAddLayer: (layer) => {
+    if (layer instanceof MapboxLayer) {
+      map.addLayer(layer, layer.props.insertBefore)
+    }
+  },
+  onRemoveLayer: (layer) => {
+    if (layer instanceof MapboxLayer) {
+      map.removeLayer(layer.id);
+    }
+  }
+});
+```
+
+Advantages:
+
+* Allows deck.gl and mapbox layers to be ordered (to some degree) within the deck.gl layer list
+* Callbacks are generic and could possibly be useful in other contexts
+
+Downsides:
+
+* Uses `MapboxLayer` inside deck.gl API, ideally that module should be reserved for mapbox API.
+* Relies on callbacks which is not "declarative"
+* Makes it harder to change/control mapbox layer structure, as injection of custom layers is based on deck.gl lifecycle and not directly under application control.
+
+
+
+### Proposal 2: LayerGroups
+
+Ideas:
+
+* "Reuse" the `LayerGroup` concept from the "Layer Group and Operations RFC" to enable the specification of groups of layers that can be rendered "separately".
+* Move towards an official static deck.gl rendering API where the app can add a list of groups and then control how they are rendered.
+
+
+```js
+import Deck, {LayerGroup, ScatterplotLayer} from '@deck.gl/core';
+import {MapboxLayer} from '@deck.gl/mapbox';
+
+class App extends React.Component {
+  render() {
+    return (
+      <Deck
+        width='100%'
+        height='100%'
+        initialViewState={{
+          longitude: -122.4,
+          latitude: 37.8,
+        }}
+
+        // A simple, callbackless way to make deck initialize with the gl context from a sub-component is envisioned...
+        gl='map'
```

Comment:
What's the purpose of having mixed mapbox/deck layers if they are being rendered into the same context?